mobile/mobile-core/android-process-service.ts

```diff
@@ -0,0 +1,189 @@
+
+			let matches = processIdRegex.exec(processIdInformation);
+
+			if (!matches || !matches[0]) {
+				return null;
+			}
+
+			return matches[1];
+		}).future<string>()();
+	}
+
+	private getAvailableAndroidPortsInformation(adb: Mobile.IDeviceAndroidDebugBridge, appIdentifier: string): IFuture<Mobile.IAndroidPortInformation[]> {
+		return (() => {
+			// Get ports information.
+			// Result will look like this:
+			// sl local_address rem_address   st tx_queue rx_queue tr tm->when retrnsmt  uid  timeout inode
+			// 0: 0100007F:1C23 00000000:0000 0A 00000000:00000000 00:00000000 00000000  1001       0 2111 1 e22cc000 300 0 0 2 -1
+			let tcpPorts: string[] = adb.executeShellCommand(["cat", "proc/net/tcp"]).wait().split(EOL);
+			let tcp6Ports: string[] = adb.executeShellCommand(["cat", "proc/net/tcp6"]).wait().split(EOL);
+
+			let allPorts: string[] = tcpPorts.concat(tcp6Ports);
+
+			// Get information only for the ports which are not in use - remote address is empty (only zeroes) and have host localhost.
+			let emptyAddressRegex = /^00*:0+/;
+
+			// Localhost hex representation is 0100007F or 0000000000000000FFFF00000100007F
+			let localHostHexRegex = /^00*100007F/;
+			let localHostVersionSixHexRegex = /^00*FFFF00000100007F/;
+			let availablePorts: Mobile.IAndroidPortInformation[] = _(allPorts)
+				.filter((line: string) => line.match(AndroidProcessService.ANDROID_PORT_INFORMATION_REGEX))
+				.map(this.parseAndroidPortInformation)
+				.filter((port: Mobile.IAndroidPortInformation) => port.remAddress.match(emptyAddressRegex) && (port.ipAddressHex.match(localHostHexRegex) || port.ipAddressHex.match(localHostVersionSixHexRegex)))
+				.value();
+
+			return availablePorts;
+		}).future<Mobile.IAndroidPortInformation[]>()();
+	}
+
+	private parseAndroidPortInformation(portInformationRow: string): Mobile.IAndroidPortInformation {
+		let matches = AndroidProcessService.ANDROID_PORT_INFORMATION_REGEX.exec(portInformationRow);
+		if (!matches || !matches[0]) {
+			// The input information does not match the regex.
+			return null;
+		}
+
+		// The local address will look like this IP_address:port both in hex format.
+		let localAddress = matches[2];
+		let localAddressParts = localAddress.split(":");
+
+		let hexIpAddress = localAddressParts[0];
+		let hexPort = localAddressParts[1];
+
+		// The first match is the whole row.
+		let portInformation: Mobile.IAndroidPortInformation = {
+			sl: matches[1],
+			localAddress: localAddress,
+			remAddress: matches[3],
+			st: matches[4],
+			uid: parseInt(matches[5]),
+			ipAddressHex: hexIpAddress,
+			number: parseInt(hexPort, 16),
+			numberHex: hexPort
+		};
+
+		return portInformation;
+	}
+
+	private getAdb(deviceIdentifier: string): Mobile.IDeviceAndroidDebugBridge {
+		return this.$injector.resolve(deviceAndroidDebugBridgePath.DeviceAndroidDebugBridge, { identifier: deviceIdentifier });
```

Comment:
This will return a new instance of `DeviceAndroidDebugBridge` every time the function is called. Maybe it will be a good idea to cache the instance in dictionary.